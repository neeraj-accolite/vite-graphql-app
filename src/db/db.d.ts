import { RxCollection, RxJsonSchema, RxDocument } from "rxdb"

export interface CartDataTypes {
    id: string;
    grandTotal: Money;
    items: Array<CartItem>;
}

interface Money {
    amount: number;
    formatted: string;
}

export interface CartItem {
    id: string;
    name: string;
    description: string;
    images: string;
    quantity: number;
    lineTotal: Money;
    unitTotal: Money;
}

export type CartCollectionMethod = {
    getCart: (id: string) => Promise<RxDocument<CartDataTypes>>;
}

export type CartCollection = RxCollection<CartDataTypes, T, CartCollectionMethod>;

export interface ShoppingAppDbCollections {
    carts: CartCollection
}