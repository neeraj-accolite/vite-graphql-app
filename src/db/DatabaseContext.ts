import { createContext } from "react";
import { RxDatabase } from "rxdb";
import { ShoppingAppDbCollections } from "./db";

export const DatabaseContext = createContext<RxDatabase<ShoppingAppDbCollections> | null>(null);
