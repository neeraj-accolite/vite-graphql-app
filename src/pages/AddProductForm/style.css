.box {
    background-color: black;
    height: 100vh;
    display: flex;
    align-items: center;
    justify-content: center;
}

.box a {
    padding: 15px;
    background-color: #fff;
    border-radius: 3px;
    text-decoration: none;
}

.close {
    position: absolute;
    top: 10px;
    right: 10px;
    color: #fe0606;
    font-size: 30px;
    text-decoration: none;
}
.modal {
    position: fixed;
    inset: 0;
    background: rgba(93, 92, 92, 0.7);
    display: none;
    align-items: center;
    justify-content: center;
}

.content {
    position: relative;
    background: white;
    padding: 1em 2em;
    border-radius: 4px;
    width: 60%;
}

.modal:target {
    display: flex;
}


#loader {
    border: 16px solid #f3f3f3; /* Light grey */
    border-top: 16px solid #3498db; /* Blue */
    border-radius: 50%;
    width: 50px;
    height: 50px;
    animation: spin 2s linear infinite;
  }
  
  @keyframes spin {
    0% { transform: rotate(0deg); }
    100% { transform: rotate(360deg); }
  }


#detailsBtn {
    margin: 10px;
  }
  
  #detailSection {
    margin: 10px;
  }
  
  #field {
    font-size: 17px;
    color: black;
    margin-right: 10px;
  }
  #value {
    color: gray;
    font-size: 17px;
  }
  
  
  #fieldOrder {
    font-size: 15px;
    color: black;
    font-weight: 600;
    margin-right: 10px;
  }
  #valueOrder {
    color: gray;
    font-size: 15px;
    font-style: italic;
  }

  .users {
    display: flex;
    background-color: white;
    border: 1px solid #f3f3f3; /* Light grey */
    padding: 10px;
    margin-top: 20px;
  }