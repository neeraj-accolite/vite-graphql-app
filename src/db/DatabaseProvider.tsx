import { useEffect, useState } from "react";
import { RxDatabase } from "rxdb";
import { ShoppingAppDbCollections } from "./db";
import { DatabaseContext } from "./DatabaseContext";
import { createDatabaseClient } from ".";


export const DatabaseProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {

    const [database, setDatabase] = useState<RxDatabase<ShoppingAppDbCollections> | null>(null);

    useEffect(() => {
        async function initiateDatabaseClient() {
            const db = await createDatabaseClient();
            setDatabase(db);
        }
        initiateDatabaseClient();
    }, []);

    if (!database) {
        //Database is not yet initialized. 
        return null;
    }

    return (
        <DatabaseContext.Provider value={database}>{children}</DatabaseContext.Provider>
    )

}