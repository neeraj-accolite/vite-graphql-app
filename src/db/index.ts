import { createRxDatabase } from "rxdb"
import { getRxStorageDexie } from "rxdb/plugins/storage-dexie"
import { cartCollectionMethod, cartSchema } from "./schemas/cart.schema"
import { ShoppingAppDbCollections } from "./db";
import { indexedDB, IDBKeyRange } from "fake-indexeddb";

export const createDatabaseClient = async () => {

    const db = await createRxDatabase<ShoppingAppDbCollections>({
        name: 'nutrien-shopping-cart',
        storage: process.env.NODE_ENV == 'test'
            ? getRxStorageDexie({
                indexedDB,
                IDBKeyRange
            })
            :
            getRxStorageDexie(),
        password: "StrongPassword@98",
        ignoreDuplicate: process.env.NODE_ENV == 'test'
    });

    await db.addCollections({
        carts: {
            schema: cartSchema,
            statics: cartCollectionMethod
        }
    });
    return db;
}