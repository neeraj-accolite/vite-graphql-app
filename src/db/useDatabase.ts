import { useContext } from "react";
import { DatabaseContext } from "./DatabaseContext";

export const useDatabase = () => {
    const context = useContext(DatabaseContext);
    if (!context) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context;
};