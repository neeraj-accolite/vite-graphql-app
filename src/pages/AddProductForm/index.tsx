import { useState, useEffect } from 'react';
import './style.css';
import { getOrderDetails, Order, OrderDetail } from '@acc/api';
import { useAddItem } from '../../hooks/useAddItem';

interface AddProductFormProps {
    cartId?: string;
}

export default function AddProductForm(props: AddProductFormProps) {
    const [isLoading, setLoading] = useState<boolean>(true);
    const [orders, setOrders] = useState<OrderDetail | null>(null);

    const onItemAddedToCart = () => {
        location.href = '#';
    }

    const { addItemToCart } = useAddItem(onItemAddedToCart);

    useEffect(() => {
        setTimeout(async () => {
            //set time out added to get the loading effects 
            const orderData = await getOrderDetails(2);
            setOrders(orderData);
            setLoading(false);
        }, 100);
    }, []);

    const onAddBtnClicked = (product: Order) => {
        addItemToCart(product, props.cartId);
    }

    return (
        <div id="add-cart-item" className='modal'>
            <div className='content'>
                <div>
                    <h2>Add Item in the Cart</h2>
                    <a href='#' className='close'>&times;</a>
                </div>
                <div>
                    {
                        isLoading ?
                            <div id="loader" style={{ marginTop: '30px', marginBottom: '30px' }}></div>
                            :
                            <div className='scrollBarForm'>
                                {
                                    orders?.products &&
                                    orders.products.map(order => {
                                        return (
                                            <div key={order.id}>
                                                <div className='users'>
                                                    <div id="detailSection">
                                                        <img src={order?.thumbnail} width={80} height={80} />
                                                    </div>
                                                    <div style={{ flex: 1 }}>
                                                        <div id="detailSection">
                                                            <label id="fieldOrder">Product:
                                                            </label>
                                                            <label id="valueOrder">
                                                                {order?.title}
                                                            </label>
                                                        </div>
                                                        <div id="detailSection">
                                                            <label id="fieldOrder">Quantity:</label>
                                                            <label id="valueOrder">{order?.quantity}</label>
                                                        </div>
                                                        <div id="detailSection">
                                                            <label id="fieldOrder">Price:</label>
                                                            <label id="valueOrder">{order?.price}</label>
                                                        </div>
                                                    </div>
                                                    <div>
                                                        <button className='addToCartBtn' onClick={() => onAddBtnClicked(order)}>‚®Å Add Item</button>
                                                    </div>
                                                </div>
                                            </div>
                                        )
                                    })
                                }
                            </div>
                    }
                </div>
            </div>
        </div>
    )
}