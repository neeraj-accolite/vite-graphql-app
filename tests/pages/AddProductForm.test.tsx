import { configure, fireEvent, render, waitFor } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import AddProductForm from '../../src/pages/AddProductForm';
import React from "react";
import { MockedProvider } from '@apollo/client/testing';
import { DatabaseProvider } from '../../src/db/DatabaseProvider';
import OrdersData from '../__mocks__/ordersData.json';

describe("AddProductForm Page", () => {

    configure({ testIdAttribute: 'id' })
    vi.mock("@acc/api", () => {
        return {
            getOrderDetails: (id: number) => new Promise((resolve, reject) => {
                resolve(OrdersData);
            })
        }
    });

    const AddProductFormComponent = () => (
        <MockedProvider addTypename={false}>
            <DatabaseProvider>
                <AddProductForm />
            </DatabaseProvider>
        </MockedProvider>
    )

    test("Add Product Form is rendered successfully with products list loaded in the form", async () => {
        const component = render(
            <AddProductFormComponent />
        )
        const elem = await waitFor(() => component.findByText(OrdersData.products[0].title));
        expect(elem).to.not.null;
        expect(component).toMatchSnapshot();
    });

    test("The Page is displaying loading till the Data is fetched", async () => {

        const component = render(
            <AddProductFormComponent />
        )
        //The loading should be enabled first.
        let loadingElem = await component.findByTestId('loader');
        expect(loadingElem).toBeTruthy();

        //once the mock data is loaded, the Product data should be available in DOM.
        const productElem = await waitFor(() => component.findByText(OrdersData.products[0].title));
        expect(productElem).toBeTruthy();
        //The loading should be disabled as well.
        let elem = component.queryByTestId('loader');
        expect(elem).not.toBeTruthy();

    });

    test("On Add Item to the Cart, the window location is set to # to disappear the form", async () => {
        vi.spyOn(require("@apollo/client"), "useMutation")
            .mockReturnValue([
                vi.fn(),
                {
                    loading: false,
                    error: null,
                    data: {
                        "message": "Item added to the cart successfully",
                        grandTotal: {
                            amount: 100,
                            formmated: "Rs. 100"
                        }
                    }
                }
            ])
        const component = render(
            <AddProductFormComponent />
        )
        const addItemBtn = (await component.findAllByText("‚®Å Add Item"))[0];
        expect(addItemBtn).toBeTruthy();
        fireEvent.click(addItemBtn, OrdersData.products[0]);
        expect(window.location.href).toContain("#");
    })

});