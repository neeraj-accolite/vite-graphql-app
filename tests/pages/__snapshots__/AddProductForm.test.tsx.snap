// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddProductForm Page > Add Product Form is rendered successfully with products list loaded in the form 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="modal"
        id="add-cart-item"
      >
        <div
          class="content"
        >
          <div>
            <h2>
              Add Item in the Cart
            </h2>
            <a
              class="close"
              href="#"
            >
              ×
            </a>
          </div>
          <div>
            <div
              class="scrollBarForm"
              style=""
            >
              <div>
                <div
                  class="users"
                >
                  <div
                    id="detailSection"
                  >
                    <img
                      height="80"
                      src="https://cdn.dummyjson.com/products/images/vehicle/Charger%20SXT%20RWD/thumbnail.png"
                      width="80"
                    />
                  </div>
                  <div
                    style="flex: 1;"
                  >
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Product:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        Charger SXT RWD
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Quantity:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        3
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Price:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        32999.99
                      </label>
                    </div>
                  </div>
                  <div>
                    <button
                      class="addToCartBtn"
                    >
                      ⨁ Add Item
                    </button>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="users"
                >
                  <div
                    id="detailSection"
                  >
                    <img
                      height="80"
                      src="https://cdn.dummyjson.com/products/images/laptops/Apple%20MacBook%20Pro%2014%20Inch%20Space%20Grey/thumbnail.png"
                      width="80"
                    />
                  </div>
                  <div
                    style="flex: 1;"
                  >
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Product:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        Apple MacBook Pro 14 Inch Space Grey
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Quantity:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        2
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Price:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        1999.99
                      </label>
                    </div>
                  </div>
                  <div>
                    <button
                      class="addToCartBtn"
                    >
                      ⨁ Add Item
                    </button>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="users"
                >
                  <div
                    id="detailSection"
                  >
                    <img
                      height="80"
                      src="https://cdn.dummyjson.com/products/images/womens-jewellery/Green%20Oval%20Earring/thumbnail.png"
                      width="80"
                    />
                  </div>
                  <div
                    style="flex: 1;"
                  >
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Product:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        Green Oval Earring
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Quantity:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        5
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Price:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        24.99
                      </label>
                    </div>
                  </div>
                  <div>
                    <button
                      class="addToCartBtn"
                    >
                      ⨁ Add Item
                    </button>
                  </div>
                </div>
              </div>
              <div>
                <div
                  class="users"
                >
                  <div
                    id="detailSection"
                  >
                    <img
                      height="80"
                      src="https://cdn.dummyjson.com/products/images/mobile-accessories/Apple%20Airpods/thumbnail.png"
                      width="80"
                    />
                  </div>
                  <div
                    style="flex: 1;"
                  >
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Product:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        Apple Airpods
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Quantity:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        5
                      </label>
                    </div>
                    <div
                      id="detailSection"
                    >
                      <label
                        id="fieldOrder"
                      >
                        Price:
                      </label>
                      <label
                        id="valueOrder"
                      >
                        129.99
                      </label>
                    </div>
                  </div>
                  <div>
                    <button
                      class="addToCartBtn"
                    >
                      ⨁ Add Item
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="modal"
      id="add-cart-item"
    >
      <div
        class="content"
      >
        <div>
          <h2>
            Add Item in the Cart
          </h2>
          <a
            class="close"
            href="#"
          >
            ×
          </a>
        </div>
        <div>
          <div
            class="scrollBarForm"
            style=""
          >
            <div>
              <div
                class="users"
              >
                <div
                  id="detailSection"
                >
                  <img
                    height="80"
                    src="https://cdn.dummyjson.com/products/images/vehicle/Charger%20SXT%20RWD/thumbnail.png"
                    width="80"
                  />
                </div>
                <div
                  style="flex: 1;"
                >
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Product:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      Charger SXT RWD
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Quantity:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      3
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Price:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      32999.99
                    </label>
                  </div>
                </div>
                <div>
                  <button
                    class="addToCartBtn"
                  >
                    ⨁ Add Item
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="users"
              >
                <div
                  id="detailSection"
                >
                  <img
                    height="80"
                    src="https://cdn.dummyjson.com/products/images/laptops/Apple%20MacBook%20Pro%2014%20Inch%20Space%20Grey/thumbnail.png"
                    width="80"
                  />
                </div>
                <div
                  style="flex: 1;"
                >
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Product:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      Apple MacBook Pro 14 Inch Space Grey
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Quantity:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      2
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Price:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      1999.99
                    </label>
                  </div>
                </div>
                <div>
                  <button
                    class="addToCartBtn"
                  >
                    ⨁ Add Item
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="users"
              >
                <div
                  id="detailSection"
                >
                  <img
                    height="80"
                    src="https://cdn.dummyjson.com/products/images/womens-jewellery/Green%20Oval%20Earring/thumbnail.png"
                    width="80"
                  />
                </div>
                <div
                  style="flex: 1;"
                >
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Product:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      Green Oval Earring
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Quantity:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      5
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Price:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      24.99
                    </label>
                  </div>
                </div>
                <div>
                  <button
                    class="addToCartBtn"
                  >
                    ⨁ Add Item
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="users"
              >
                <div
                  id="detailSection"
                >
                  <img
                    height="80"
                    src="https://cdn.dummyjson.com/products/images/mobile-accessories/Apple%20Airpods/thumbnail.png"
                    width="80"
                  />
                </div>
                <div
                  style="flex: 1;"
                >
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Product:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      Apple Airpods
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Quantity:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      5
                    </label>
                  </div>
                  <div
                    id="detailSection"
                  >
                    <label
                      id="fieldOrder"
                    >
                      Price:
                    </label>
                    <label
                      id="valueOrder"
                    >
                      129.99
                    </label>
                  </div>
                </div>
                <div>
                  <button
                    class="addToCartBtn"
                  >
                    ⨁ Add Item
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
